### PR TITLE
Frontend Challenge

### PR DESCRIPTION
**Pull request by vmontala**
Hi there, he'll you find my solved version of your exercise.

It has been much harder than I initially though, but it was fun enough 😄 

There are a couple of comments regarding the app:
- ~I haven't included any graph library as I did not know which data could I show (all items on the long list have the same details), I also did not want to delay more the review~ (going to add it soon)
- ~Though it was asked to include CSS animations, I included a library that handles them (to not spend more time and not reinvent the wheel)~ (will be done on graphs)
- The code may look a bit messy, I would liked to have one single model instead of two, but as the data is that different and the components are so, I thought it was better this way
- I tried to include the fonts locally with Webpack, but after a couple of tries I gave up and imported them from Google
- If I knew which would be the next steps on the application, or if the application is going to be encapsulated inside a larger one, I'd split some stuff (like buttons, etc) as components, to make the code more reusable
- Responsive design is not perfect, though I'm quite happy with it, there was lots of data to show there and I did not thought about any better solution

### CODE REVIEW

**Comment by vmontala** [comment]
I'm allowing myself to do some small cleanup 😄 ✂️

**Comment by vmontala** [comment]
@cavalheiro I'm waiting for @mateiw to accept the collaboration request 😃

Btw, I thought about adding a new feature to randomise the list data when being modeled, that way it will be different enough to have some graphs, I hope it's okay if I add it this afternoon ✌️

File: app/src/App/index.js [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    this.toggleListLayout = this.toggleListLayout.bind(this)
+  }
+
+  /**
+   * Fetches the machine list from the server and stores the list of machines in the state
+   */
+  fetchMachines() {
+    Machines.getAll()
+      .then((machines) => this.setState({ machines }))
+      .finally(() => this.setState({ overlaySpinner: false }))
```
Comment (mateiw) [inline app/src/App/index.js]: Some err handling would be needed here.

File: app/src/App/index.js [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+      machine: null,
+      currentPage: 1,
+      pageLimit: 5,
+      isExtendedList: true,
+      filter: '',
+      overlaySpinner: true,
+      detailSpinner: false
+    }
+
+    this.fetchMachines()
```
Comment (mateiw) [inline app/src/App/index.js]: I think async calls should be done from `componendDidUpdate`.

File: app/src/App/index.js [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  /**
+   * Given an id, it fetches a machine detail from the server and stores it in the state
+   *
+   * @param {number} id - ID from the machine to fetch
+   */
+  fetchMachine(id) {
+    Machines.getById(id)
+      .then((machine => this.setMachine(machine)))
```
Comment (mateiw) [inline app/src/App/index.js]: Here too some err handling would be needed.

File: app/src/Pagination/index.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+/**
+ * Machine list pagination component, it renders the pagination, handles the button status and
+ * contains the navigation logic
+ */
+class Pagination extends Component {
+  constructor() {
+    super()
+
+    this.setFirstPage = this.setFirstPage.bind(this)
```
Comment (mateiw) [inline app/src/Pagination/index.js]: You can also use arrow functions with property initializers to bind the function: `method = () => { ... }`.

File: app/src/App/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (vmontala) [reply app/src/App/index.js]: Right, I did it this way on the very beginning and I forgot to change it, nice catch!

File: app/src/Pagination/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (vmontala) [reply app/src/Pagination/index.js]: Haven't checked that much ES7, I'll take a look after all the fixes

File: app/src/App/index.js (hunk 2 above)
Comment (vmontala) [reply app/src/App/index.js]: Adding an async call that modifies the state in the `componentDidUpdate` would end in an infinite loop, as after modifying the state the component is updated, it'd trigger that function, requesting the data again, and starting the cycle again...

Also, `componentDidUpdate` is not triggered on the first load of the component, and we want the data by then, right?

File: app/src/App/index.js (hunk 1 above)
Comment (vmontala) [reply app/src/App/index.js]: I thought when adding the spinner and I forgot, adding it now 😃

File: app/src/App/index.js (hunk 3 above)
Comment (vmontala) [reply app/src/App/index.js]: I thought when adding the spinner and I forgot, adding it now 😃

**Comment by vmontala** [comment]
@mateiw thanks for the feedback!

I did the changes you suggested (and some more, mostly refactors), looking forward for some more comments 😄

**Comment by vmontala** [comment]
Hi @ncounter!

Thanks for your feedback! I'm glad you liked it so far.

About your two questions;
- I'd say that for a higher overview of the systems the best solution would be a dashboard-like page with all relevant information.
That depends on the requisites (e.g. if it's not normal to have not running machines, display the total/small list of stopped machines, so they can be seen at first glance) also, I'd probably include some graphs as they work better for comparing data (e.g. if we not just want to know if there are any machine malfunctioning, but also to compare it to the total of machines)
- Regarding the color-blind users, I think that the solution is to include icons to represent statuses, such as active/not active, enabled/disabled auto-update, etc

Fun fact; if graphs were included, what about the color-blind users? I know that there are solutions such as coloured backgrounds with patterns so they can be identified easier, but I have never talked with a color-blind person about that 😄

**Pull request closed without merging**